flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManagerImpl.java

```diff
@@ -807,23 +811,32 @@ private boolean isFulfillableByRegisteredSlots(ResourceProfile resourceProfile)
 		return false;
 	}
 
-	private Optional<PendingTaskManagerSlot> allocateResource(ResourceProfile resourceProfile) {
-		final Collection<ResourceProfile> requestedSlots = resourceActions.allocateResource(resourceProfile);
+	private Optional<PendingTaskManagerSlot> allocateResource(ResourceProfile requestedSlotResourceProfile) {
+		if (defaultWorkerResourceSpec == null) {
+			// standalone mode, cannot allocate resource
+			return Optional.empty();
+		}
 
-		if (requestedSlots.isEmpty()) {
+		if (!Preconditions.checkNotNull(defaultSlotResourceProfile,
+			"defaultSlotResourceProfile should be null iff taskExecutorProcessSpec is null, which means standalone mode.")
+			.isMatching(requestedSlotResourceProfile)) {
+			// requested resource profile is unfulfillable
 			return Optional.empty();
-		} else {
-			final Iterator<ResourceProfile> slotIterator = requestedSlots.iterator();
-			final PendingTaskManagerSlot pendingTaskManagerSlot = new PendingTaskManagerSlot(slotIterator.next());
-			pendingSlots.put(pendingTaskManagerSlot.getTaskManagerSlotId(), pendingTaskManagerSlot);
+		}
```

Comment:
Shouldn't we instead of returning an `Optional.empty()` fail with an invalid state exception because the current `SlotManagerImpl` does not support custom resource profiles yet?

Reply:
I'm not sure whether this should be an invalid state exception. 

It is true that all the slot requests should have unknown resource profile ATM. However, I think `SlotManagerImpl` does not necessarily need to know whether there're custom profiles or not. As long as the requested resource profile can be matched by the default slot profile, it should be able to allocate the slot. And if the requested slot profile is larger than the default slot, an `UnfulfillableSlotRequestException` should be thrown.

Reply:
Failing in case that the requested slot profile is too large sound good to me.